File: client/src/main/java/singh/harneev/atcclient/MainActivity.java
package singh.harneev.atcclient;

import android.app.Service;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import java.util.List;

import singh.harneev.atcsdk.IDBCallback;
import singh.harneev.atcsdk.IDBService;
import singh.harneev.atcsdk.school.School;

public class MainActivity extends AppCompatActivity {

    private static final String IPC_SERVICE_ACTION = "singh.harneev.atc.IDBService";
    private static final String IPC_SERVICE_PACKAGE = "singh.harneev.atc";

    private IDBService mIPCService;
    private ServiceConnection mServiceConnection;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    protected void onResume() {
        super.onResume();
        // binding to  IPC services
        initConnection();
    }

    public void showAllSchoolList(View view) {

        try {
            mIPCService.getSchoolDetails(serviceCallback);
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public void SearchSchoolInfo(View view) {

    }

    IDBCallback.Stub serviceCallback = new IDBCallback.Stub() {

        @Override
        public void handleSchoolDetail(final List<School> schoolList) throws RemoteException {

            runOnUiThread(new Runnable() {
                @Override
                public void run() {

                    Toast.makeText(MainActivity.this, "count: "+ schoolList.size(), Toast.LENGTH_LONG).show();
                }
            });
        }
    };

    /**
     * Bind to IPC Services
     */
    private void initConnection() {

        mServiceConnection = new ServiceConnection() {

            @Override
            public void onServiceDisconnected(ComponentName name) {
                mIPCService = null;
                Toast.makeText(getApplicationContext(), "Service Disconnected",
                        Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onServiceConnected(ComponentName name, IBinder service) {
                try {
                    mIPCService = IDBService.Stub.asInterface((IBinder) service);

                    Toast.makeText(getApplicationContext(), "Service Connected",
                            Toast.LENGTH_SHORT).show();

                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };

        // First time initialization
        if (mIPCService == null) {
            Intent it = new Intent();
            it.setAction(IPC_SERVICE_ACTION);
            it.setPackage(IPC_SERVICE_PACKAGE);

            // binding to remote service
            bindService(it, mServiceConnection, Service.BIND_AUTO_CREATE);
        }
    }
}

File: client/src/main/java/singh/harneev/atcclient/MainActivity.kt
package singh.harneev.atcclient

import singh.harneev.atcsdk.IDBService
import android.content.ServiceConnection
import android.os.Bundle
import singh.harneev.atcclient.R
import singh.harneev.atcsdk.IDBCallback
import kotlin.Throws
import singh.harneev.atcsdk.school.School
import android.widget.Toast
import android.content.ComponentName
import android.os.IBinder
import android.content.Intent
import android.os.RemoteException
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import java.lang.Exception

class MainActivity : AppCompatActivity() {
    private var mIPCService: IDBService? = null

    private var mServiceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(name: ComponentName) {
            mIPCService = null
            Toast.makeText(
                applicationContext, "Service Disconnected",
                Toast.LENGTH_SHORT
            ).show()
        }

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            mIPCService = IDBService.Stub.asInterface(service)
            Toast.makeText(
                applicationContext, "Service Connected",
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    private val serviceCallback: IDBCallback.Stub = object : IDBCallback.Stub() {
        @Throws(RemoteException::class)
        override fun handleSchoolDetail(schoolList: List<School>) {
            runOnUiThread {
                Toast.makeText(
                    this@MainActivity,
                    "count: " + schoolList.size,
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onResume() {
        super.onResume()
        // binding to  IPC services

        mIPCService ?: run {
            val intent = Intent(IPC_SERVICE_ACTION).also {
                it.`package` = IPC_SERVICE_PACKAGE
            }

            bindService(intent, mServiceConnection, BIND_AUTO_CREATE)
        }
    }

    fun showAllSchoolList(view: View?) {
        try {
            mIPCService?.getSchoolDetails(serviceCallback)
        } catch (e: RemoteException) {
            e.printStackTrace()
        }
    }

    fun SearchSchoolInfo(view: View?) {}

    companion object {
        private const val IPC_SERVICE_ACTION = "singh.harneev.atc.IDBService"
        private const val IPC_SERVICE_PACKAGE = "singh.harneev.atc"
    }
}